### PR TITLE
sci-mathematics/octave: bump to version 9.2.0

### PR DESCRIPTION
**Pull request by efferre79**
I have added support to Qt6. The IUSE=gui might be removed but I left it to avoid further modifications (e.g. change of mask file for hppa)

Please check all the boxes that apply:

- [x] I can submit this contribution in agreement with the [Copyright Policy](https://www.gentoo.org/glep/glep-0076.html#certificate-of-origin).
- [x] I have certified the above via adding a `Signed-off-by` line to *every* commit in the pull request.
- [x] This contribution has not been created with the assistance of Natural Language Processing artificial intelligence tools, in accordance with the [AI policy](https://wiki.gentoo.org/wiki/Project:Council/AI_policy).
- [x] I have run `pkgcheck scan --commits --net` to check for issues with my commits.

Please note that all boxes must be checked for the pull request to be merged.

### CODE REVIEW

File: sci-mathematics/octave/octave-9.1.0.ebuild [hunk 1]
```diff
@@ -0,0 +1,273 @@
...
+# Building without either USE=gnuplot or USE=opengl is technically legal,
+# but will leave you unable to plot anything.
+#
+# Octave's FLTK support is unofficially deprecated, in the sense that
+# you'll often get "why are you using FLTK?" in response to
+# bugs. (Upstream bug 59321 for a random example.) In the future, it
+# will probably make sense to merge USE=opengl and USE=X into USE=qt[56],
+# dropping USE=fltk entirely.
+REQUIRED_USE="
+	?? ( qt5 qt6 )
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.1.0.ebuild:136]: Per [Qt packaging policy](https://wiki.gentoo.org/wiki/Project:Qt/Policies) we should not introduce the `qt5` flag and `REQUIRED_USE` here.

Instead we should add only the `qt6` flag, if enabled gui is built with qt6, otherwise it is built with qt5.

Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.1.0.ebuild:136]:
```suggestion
```

File: sci-mathematics/octave/octave-9.1.0.ebuild [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+		x11-libs/gl2ps:=
+	)
+	portaudio? ( media-libs/portaudio )
+	postscript? (
+		app-text/epstool
+		media-gfx/pstoedit
+		>=media-gfx/fig2dev-3.2.9-r1
+	)
+	gui? (
+		qt5? (
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.1.0.ebuild:63]:
```suggestion
		!qt6? (
```

File: sci-mathematics/octave/octave-9.1.0.ebuild [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+DESCRIPTION="High-level interactive language for numerical computations"
+HOMEPAGE="https://www.gnu.org/software/octave/"
+SRC_URI="mirror://gnu/${PN}/${P}.tar.xz"
+
+LICENSE="GPL-3"
+SLOT="0/${PV}"
+KEYWORDS="~amd64 ~arm ~arm64 ~hppa ~ppc ~ppc64 ~riscv ~x86 ~amd64-linux ~x86-linux"
+
+IUSE="curl doc fftw fltk +glpk gnuplot gui hdf5 imagemagick java json klu opengl portaudio postscript +qhull +qrupdate +qt5 qt6 readline sndfile +sparse spqr ssl sundials X zlib"
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.1.0.ebuild:16]:
```suggestion
IUSE="curl doc fftw fltk +glpk gnuplot gui hdf5 imagemagick java json klu opengl portaudio postscript +qhull +qrupdate qt6 readline sndfile +sparse spqr ssl sundials X zlib"
```

File: sci-mathematics/octave/octave-9.1.0.ebuild [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+# "makeinfo" from sys-apps/texinfo at runtime to convert its texinfo
+# documentation to text (see scripts/help/help.m).
+#
+# (un)zip isn't mentioned, but there's a test that uses it (bug #775254).
+#
+# The use of USE=imagemagick to pull in media-gfx/graphicsmagick is not
+# ideal, but both "graphicsmagick" and "imagemagick" are global USE
+# flags whose existing descriptions conflict with the obvious way we
+# would want to use them in octave. In any case, upstream doesn't really
+# support imagemagick, only graphicsmagick (bug 864785).
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.1.0.ebuild:28]: Is this still the case? I find the graphicsmagick/imagemagick situation here a bit confusing.

File: sci-mathematics/octave/octave-9.1.0.ebuild [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+# you'll often get "why are you using FLTK?" in response to
+# bugs. (Upstream bug 59321 for a random example.) In the future, it
+# will probably make sense to merge USE=opengl and USE=X into USE=qt[56],
+# dropping USE=fltk entirely.
+REQUIRED_USE="
+	?? ( qt5 qt6 )
+	fltk? ( opengl X )
+	gui? ( X )
+	qt5? ( gui )
+	qt6? ( gui )
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.1.0.ebuild:140]:
```suggestion
```

File: sci-mathematics/octave/octave-9.1.0.ebuild [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+	elif use qt6 ; then
+		export MOC="$(qt6_get_bindir)/../libexec/moc" \
+			UIC="$(qt6_get_bindir)/../libexec/uic" \
+			RCC="$(qt6_get_bindir)/../libexec/rcc" \
+			LRELEASE="$(qt6_get_bindir)/lrelease" \
+			QHELPGENERATOR="$(qt6_get_bindir)/../libexec/qhelpgenerator"
+		gui_with="--with-qt=6"
+	else
+		gui_with="--without-qt"
+	fi
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.1.0.ebuild:185]:
```suggestion
	if use gui; then
		if use qt6; then
			export MOC="$(qt6_get_bindir)/../libexec/moc" \
			UIC="$(qt6_get_bindir)/../libexec/uic" \
			RCC="$(qt6_get_bindir)/../libexec/rcc" \
			LRELEASE="$(qt6_get_bindir)/lrelease" \
			QHELPGENERATOR="$(qt6_get_bindir)/../libexec/qhelpgenerator"
			gui_with="--with-qt=6"
		else
			export MOC="$(qt5_get_bindir)/moc" \
			UIC="$(qt5_get_bindir)/uic" \
			RCC="$(qt5_get_bindir)/rcc" \
			LRELEASE="$(qt5_get_bindir)/lrelease" \
			QCOLLECTIONGENERATOR="$(qt5_get_bindir)/qcollectiongenerator" \
			QHELPGENERATOR="$(qt5_get_bindir)/qhelpgenerator"
			gui_with="--with-qt=5"
		fi
	else
		gui_with="--without-qt"
	fi
```

File: sci-mathematics/octave/octave-9.1.0.ebuild [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+		--disable-64
+		--enable-shared
+		--with-z
+		--with-bz2
+
+		# bug #901965
+		--without-libiconv-prefix
+		--without-libreadline-prefix
+
+		${gui_with}
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.1.0.ebuild:208]: Instead of using a variable, lets put the qt conditionals below the definition of `myeconfargs` and then we can do `myeconfargs+=( --with(out)-qt... )` in the qt conditional block.

**Review by AndrewAmmerlaan — COMMENTED** [review]
Thanks for the PR, sorry for the delay. This looks mostly good to me, but it needs some changes to adhere to the Qt packaging policy.

And I see that in the meantime there is already a 9.2 release, so lets skip 9.1 and go immediately for 9.2.

File: sci-mathematics/octave/octave-9.1.0.ebuild (hunk 4 above)
Comment (orlitzky) [reply sci-mathematics/octave/octave-9.1.0.ebuild:28]: AFAIK. GraphicsMagick has stable API and is careful about ABI, while ImageMagick does whatever whenever. GM is much nicer to have as a dependency.

**Comment by orlitzky** [comment]
FLTK is now officially deprecated, so I think it's time to drop the option. I would also strongly recommend supporting only qt6. It's stable, and the 9.2.0 release notes say that qt5 support may be dropped.

tl;dr I would just put qt6 behind USE=gui and delete the rest of the gui stuff. There is a big comment above `REQUIRED_USE` explaining the GUI mess that could then be deleted.

**Comment by efferre79** [comment]
I have pushed all suggested changes

File: sci-mathematics/octave/octave-9.2.0.ebuild [hunk 8]
```diff
@@ -110,36 +97,13 @@ BDEPEND="
 		dev-texlive/texlive-metapost
 		virtual/latex-base
 	)
-	qt5? ( dev-qt/linguist-tools:5 )
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.2.0.ebuild]: This one we should probably replace with `gui? ( dev-qt/qttools:6[linguist )` (unless of course qt linguist is no longer required in the qt6 version)

File: sci-mathematics/octave/octave-9.2.0.ebuild [hunk 9]
```diff
@@ -110,36 +97,13 @@ BDEPEND="
...
-# is all that FLTK is used for, so it doesn't make sense to enable
-# USE=fltk without USE=opengl.
-#
-# Building without either USE=gnuplot or USE=opengl is technically legal,
-# but will leave you unable to plot anything.
-#
-# Octave's FLTK support is unofficially deprecated, in the sense that
-# you'll often get "why are you using FLTK?" in response to
-# bugs. (Upstream bug 59321 for a random example.) In the future, it
-# will probably make sense to merge USE=opengl and USE=X into USE=qt[56],
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.2.0.ebuild]: Let's drop the opengl and X flags as suggested here. So enabling `USE=gui` would also enable the opengl and X configure options and dependencies.

File: sci-mathematics/octave/octave-9.2.0.ebuild [hunk 10]
```diff
@@ -239,6 +180,18 @@ src_configure() {
...
 	)
 
+	# Tell autoconf where to find qt binaries, fix bug #837752
+	if use gui ; then
+		export MOC="$(qt6_get_bindir)/../libexec/moc" \
+			UIC="$(qt6_get_bindir)/../libexec/uic" \
+			RCC="$(qt6_get_bindir)/../libexec/rcc" \
+			LRELEASE="$(qt6_get_bindir)/lrelease" \
+			QHELPGENERATOR="$(qt6_get_bindir)/../libexec/qhelpgenerator"
+		myeconfargs+=( "--with-qt=6" )
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.2.0.ebuild:190]: and `--with-x` and `--with-opengl`

File: sci-mathematics/octave/octave-9.2.0.ebuild [hunk 11]
```diff
@@ -239,6 +180,18 @@ src_configure() {
...
+	else
+		myeconfargs+=( "--without-qt" )
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.2.0.ebuild:192]: and `--without-x` and `--without-opengl`

File: sci-mathematics/octave/octave-9.2.0.ebuild [hunk 12]
```diff
@@ -59,20 +58,8 @@ COMMON_DEPS="
...
-			dev-qt/qtprintsupport:5
-			dev-qt/qtwidgets:5
-			x11-libs/qscintilla:=[qt5]
-		)
-		qt6? (
-			dev-qt/qtbase:6[gui,opengl,network,widgets]
-			x11-libs/qscintilla:=[qt6]
-		)
+		dev-qt/qtbase:6[gui,opengl,network,widgets]
+		x11-libs/qscintilla:=[qt6]
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.2.0.ebuild:62]: and deps currently behind `X?` and `opengl?`

**Review by AndrewAmmerlaan — COMMENTED** [review]
Thanks!

Since we are now dropping `USE=fltk`, I think we should follow the suggestion from the comment above `REQUIRED_USE` and also merge `opengl` and `X` into `gui`.

> In the future, it will probably make sense to merge USE=opengl and USE=X into USE=qt[56], dropping USE=fltk entirely.

**Comment by AndrewAmmerlaan** [comment]
> isn't `${PN}/${PV}` equal to `${P}`?

No `${P}` == `${PN}-${PV}`

File: sci-mathematics/octave/octave-9.2.0.ebuild (hunk 8 above)
Comment (efferre79) [reply sci-mathematics/octave/octave-9.2.0.ebuild]: it compiles without that

File: sci-mathematics/octave/octave-9.1.0.ebuild (hunk 4 above)
Comment (efferre79) [reply sci-mathematics/octave/octave-9.1.0.ebuild]: https://bugs.gentoo.org/864785#c13 as documented by you: I would keep current configuration (i.e. only graphicsmagick). It's really confusing the USE=imagemagick but I guess that is out of the scope of the current PR

File: sci-mathematics/octave/octave-9.2.0.ebuild (hunk 8 above)
Comment (AndrewAmmerlaan) [reply sci-mathematics/octave/octave-9.2.0.ebuild]: > uhm, actually I have already installed `dev-qt/qttools:6[linguist]` on my system so I am not sure

Lets just keep it in then, just to be sure. I can test later if it is actually used.

File: sci-mathematics/octave/octave-9.1.0.ebuild (hunk 4 above)
Comment (orlitzky) [reply sci-mathematics/octave/octave-9.1.0.ebuild]: IIRC I went with `USE=imagemagick` because the description of `USE=graphicsmagick` is,

> graphicsmagick - Build and link against GraphicsMagick instead of ImageMagick (requires USE=imagemagick if optional)

and I thought it would be pretty silly to have both flags and then the necessary `REQUIRED_USE="imagemagick? ( graphicsmagick )"` to force the right implementation.

File: sci-mathematics/octave/octave-9.2.0.ebuild [hunk 13]
```diff
@@ -0,0 +1,227 @@
...
+
+	# Tell autoconf where to find qt binaries, fix bug #837752
+	if use gui ; then
+		export MOC="$(qt6_get_bindir)/../libexec/moc" \
+			UIC="$(qt6_get_bindir)/../libexec/uic" \
+			RCC="$(qt6_get_bindir)/../libexec/rcc" \
+			LRELEASE="$(qt6_get_bindir)/lrelease" \
+			QHELPGENERATOR="$(qt6_get_bindir)/../libexec/qhelpgenerator"
+		myeconfargs+=(
+			"--with-opengl"
```
Comment (orlitzky) [inline sci-mathematics/octave/octave-9.2.0.ebuild:183]: I think you can still use `use_with` here to avoid repetition: `use_with gui opengl`, `use_with gui x`, `use_with gui qt 6`.

File: sci-mathematics/octave/octave-9.2.0.ebuild [hunk 14]
```diff
@@ -0,0 +1,227 @@
...
+		virtual/latex-base
+	)
+	gui? ( dev-qt/qttools:6[linguist] )
+	qrupdate? ( app-misc/pax-utils )
+	sparse? ( app-misc/pax-utils )
+"
+
+PATCHES=(
+	"${FILESDIR}"/${PN}-5.1.0-pkgbuilddir.patch
+	"${FILESDIR}"/${PN}-9.2.0-omit-qtchooser-qtver.patch
```
Comment (orlitzky) [inline sci-mathematics/octave/octave-9.2.0.ebuild:105]: Is the patch still needed when only qt:6 is supported?

Comment (efferre79) [reply sci-mathematics/octave/octave-9.2.0.ebuild]: indeed it is not needed anymore

File: sci-mathematics/octave/octave-9.2.0.ebuild [hunk 15]
```diff
@@ -180,15 +179,15 @@ src_configure() {
 			LRELEASE="$(qt6_get_bindir)/lrelease" \
 			QHELPGENERATOR="$(qt6_get_bindir)/../libexec/qhelpgenerator"
 		myeconfargs+=(
-			"--with-opengl"
+			$(use_with gui opengl)
+			$(use_with gui x)
 			"--with-qt=6"
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.2.0.ebuild]: I think what @orlitzky meant here is to put it higher up, outside of this if, so we don't have to write it twice.

File: sci-mathematics/octave/octave-9.2.0.ebuild [hunk 16]
```diff
@@ -0,0 +1,226 @@
...
+	if use gui ; then
+		export MOC="$(qt6_get_bindir)/../libexec/moc" \
+			UIC="$(qt6_get_bindir)/../libexec/uic" \
+			RCC="$(qt6_get_bindir)/../libexec/rcc" \
+			LRELEASE="$(qt6_get_bindir)/lrelease" \
+			QHELPGENERATOR="$(qt6_get_bindir)/../libexec/qhelpgenerator"
+		myeconfargs+=(
+			$(use_with gui opengl)
+			$(use_with gui x)
+			"--with-qt=6"
```
Comment (AndrewAmmerlaan) [inline sci-mathematics/octave/octave-9.2.0.ebuild:184]: `"--with-qt=6"` and this can be `$(use_with gui qt 6).

According to [1], the third argument will be put after the `=`.


[1] https://devmanual.gentoo.org/function-reference/query-functions/index.html

**Comment by AndrewAmmerlaan** [comment]
Thanks :+1: 

Merged in: https://gitweb.gentoo.org/repo/gentoo.git/commit/?id=f41f655d60bca2e2d3d878a03836d409ab5f8d9c
